[#1] Github Action을 활용하여 자동화된 CI/CD 워크플로우를 구성한다.

y00njinuk opened the pull request:
## Goals
- Github Action을 활용하여 워크플로우를 구성한다.

## Implements
- [x] workflows 파일 추가
- [x] build.gradle 수정

## Related Issue 
- https://github.com/jaemisam-a/dangol-sonnim-backend/issues/1

## Test
- N/A

y00njinuk committed:
feat: add dockerfile and docker-compose.yml

- AWS EC2에서 Docker에 배포할 이미지에 대한 스펙 및 커맨드 정의

y00njinuk committed:
feat: define & add Github Action Workflow for CI/CD

- CI/CD 용도의 Github Action Workflow 정의

y00njinuk committed:
feat: add branch on push events of workflow

- Github Action Workflow가 동작하는 브랜치 범위 추가

y00njinuk committed:
build: rename of branch (master->main) and remove feature branch

y00njinuk committed:
build: downgrade jdk and spring boot version

- build 설정에 JDK 17로 호환이 되는 부분 삭제
- Spring boot minor 버전 변경

y00njinuk commented:
오프라인 리뷰를 진행하였으므로 코드 리뷰 내용은 생략을 하도록 하겠습니다.
코드에 대한 질문이 있으시면 코멘트 부탁드리겠습니다.

rnjstmdals6 commented:
YML_PROD 깃 시크릿 파일을 하나 만들고

빌드 이전 리소스 폴더에 application-prod.yml 파일을 작성하는 과정을 넣어주시면 좋을 것 같습니다!

y00njinuk commented:
> YML_PROD 깃 시크릿 파일을 하나 만들고
> 
> 빌드 이전 리소스 폴더에 application-prod.yml 파일을 작성하는 과정을 넣어주시면 좋을 것 같습니다!

혹시, 이 부분에 대하여 추가 설명이 가능하실까요? 해당 작업이 왜 필요한지 구체적으로 떠오르지가 않아서요.

y00njinuk commented:
컥.. @Hyeonziii 님. approve 해주셨는데 저가 모르고 다시 요청을 드렸네요... ㅎㅎ 문제없으면 재승인 부탁드립니다.

rnjstmdals6 commented:
@y00njinuk 
application.properties(application.yml) 은 리소스 폴더에 존재하며
스프링부트가 애플리케이션을 구동할 때 자동으로 로딩하는 파일입니다.
이 파일에는 애플리케이션에서 사용하는 여러가지 설정 값과 보안적으로 중요한 값들을 저장해놓습니다. 
db 루트 비밀번호, jwt 시크릿 키 등등..
그래서 스프링부트 구동 시 꼭 필요한 파일이지만 외부에 노출되면 안되기에 git ignore에 필수로 등록하는 파일입니다.

git ignore 에 application.yml 등록하면 깃 레포지토리에 올라가지 않게 되고
결국 깃 액션에서는 application.yml 없이 서버를 배포하게 됩니다.
설정 값 없이 서버를 돌릴 수는 없기 때문에 문제가 생깁니다!

그래서 따로 깃 시크릿에 application.yml 을 저장할 수 있는 파일을 생성하고
깃 액션 배포 시 깃 시크릿에 존재하는 해당 파일을 읽어와 리소스 폴더에 저장할 수 있도록 해야할 것 같습니다!

rnjstmdals6 commented on dockerfile:
```diff
@@ -0,0 +1,19 @@
+# STEP1) Docker에 해당 이미지를 실행시킬 때 jdk17 버전으로 세팅
+FROM openjdk:17
```
자바 버전 11로 교체가 필요해보입니다!

y00njinuk replied:
수정해놓도록 하겠습니다.

rnjstmdals6 commented on build.gradle:
```diff
@@ -1,12 +1,11 @@
 
 group = 'com.dangol'
 version = '0.0.1-SNAPSHOT'
-sourceCompatibility = '17'
```
sourceCompatibility 를 11로 바꾸지 않고 지우신 이유가 궁금합니다!

y00njinuk replied:
컥.. 기본값이 11인줄 알고  추가를 해두지 않았거든요. 추가해놓도록 하겠습니다.

rnjstmdals6 replied:
감사합니다 :)

y00njinuk commented:
> @y00njinuk application.properties(application.yml) 은 리소스 폴더에 존재하며 스프링부트가 애플리케이션을 구동할 때 자동으로 로딩하는 파일입니다. 이 파일에는 애플리케이션에서 사용하는 여러가지 설정 값과 보안적으로 중요한 값들을 저장해놓습니다. db 루트 비밀번호, jwt 시크릿 키 등등.. 그래서 스프링부트 구동 시 꼭 필요한 파일이지만 외부에 노출되면 안되기에 git ignore에 필수로 등록하는 파일입니다.
> 
> git ignore 에 application.yml 등록하면 깃 레포지토리에 올라가지 않게 되고 결국 깃 액션에서는 application.yml 없이 서버를 배포하게 됩니다. 설정 값 없이 서버를 돌릴 수는 없기 때문에 문제가 생깁니다!
> 
> 그래서 따로 깃 시크릿에 application.yml 을 저장할 수 있는 파일을 생성하고 깃 액션 배포 시 깃 시크릿에 존재하는 해당 파일을 읽어와 리소스 폴더에 저장할 수 있도록 해야할 것 같습니다!

설명 감사합니다. 우선은 applycation.yaml 파일의 현재 내용으로만 구성한다고 생각하고 action을 하나 추가해두도록 하지요.

y00njinuk committed:
chore: remove unused properties

- 사용하지 않는 application.properties 파일 삭제

y00njinuk committed:
build: add application.yml to .gitignore

- 보안과 관련된 설정값들을 마스킹 처리하기 위해 github에 업로드하지 않도록 ignore 처리

y00njinuk committed:
build: write properties to application.yml

- Repository에 설정된 변수값을 바탕으로 application.yml 작성 추가
- https://github.com/jaemisam-a/dangol-sonnim-backend/settings/variables/actions 참고

y00njinuk committed:
style: renamed from cicd.yml to master-cicd.yml

y00njinuk committed:
feat: define & add Github Action Workflow for Develop

- 개발 테스트 용도의 Github Action Workflow 정의

y00njinuk commented on .gitignore:
```diff
@@ -4,6 +4,7 @@ build/
 !gradle/wrapper/gradle-wrapper.jar
 !**/src/main/**/build/
 !**/src/test/**/build/
+src/main/resources/application.yml
```
application.yml 파일은 ignore 처리합니다.

y00njinuk commented on .github/workflows/cicd.yml:
```diff
@@ -39,32 +39,42 @@ jobs:
+
+      - name: STEP3) Write to application.yml
+        run: echo "${{ vars.APP_PROPERTIES }}" > ./src/main/resources/application.yml
+        shell: bash
```
STEP2와 STEP3 과정을 거쳐서 APP_PROPERTIES에 작성된 설정값들을 application.yml 파일에 작성할 수 있도록 Github Action을 추가하였습니다.

y00njinuk commented on .github/workflows/develop-ci.yml:
```diff
@@ -0,0 +1,83 @@
+# Reference
+# https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
+
+name: dangol-sonnim-backend Develop Workflow
```
개발 테스트 용도의 workflow을 생성하였습니다. 해당 워크플로우에서는 gradle test와 build을 수행함으로써 향후 작성될 소스에 대한 컴파일과 구성된 MySQL 환경을 바탕으로 Integration Test를 할 수 있도록 하였습니다.
(+ 테스트 하기 위해서는 Workflow에 테이블 삽입하는 내용도 나중에 추가가 되어야 할 것 같네요.)

rnjstmdals6 reviewed (approved):
고생하셨습니다!

rnjstmdals6 commented:
고생하셨습니다
